internal/evaluator/resolver.go

```diff
@@ -70,39 +69,25 @@ func resolveExt[D protoreflect.Descriptor, C proto.Message](
 	desc D,
 	extType protoreflect.ExtensionType,
 ) (constraints C) {
-	opts := desc.Options().ProtoReflect()
-	fDesc := extType.TypeDescriptor()
-
-	if opts.Has(fDesc) {
-		msg := opts.Get(fDesc).Message().Interface()
-		if m, ok := msg.(C); ok {
-			return m
+	num := extType.TypeDescriptor().Number()
+	var msg proto.Message
+	proto.RangeExtensions(desc.Options(), func(typ protoreflect.ExtensionType, i interface{}) bool {
+		if num != typ.TypeDescriptor().Number() {
+			return true
 		}
-	}
-
-	unknown := opts.GetUnknown()
-	if len(unknown) == 0 {
-		return constraints
-	}
+		msg, _ = i.(proto.Message)
+		return false
+	})
 
-	opts = opts.Type().New()
-	var resolver protoregistry.Types
-	if err := resolver.RegisterExtension(extType); err != nil {
+	if msg == nil {
 		return constraints
-	}
-	_ = proto.UnmarshalOptions{Resolver: &resolver}.Unmarshal(unknown, opts.Interface())
-
-	if opts.Has(fDesc) {
-		msg := opts.Get(fDesc).Message().Interface()
-		if m, ok := msg.(C); ok {
-			return m
-		}
-	}
-	msg := opts.Get(fDesc).Message().Interface()
-	if m, ok := msg.(C); ok {
+	} else if m, ok := msg.(C); ok {
 		return m
 	}
 
+	constraints, _ = constraints.ProtoReflect().New().Interface().(C)
+	b, _ := proto.Marshal(msg)
+	_ = proto.Unmarshal(b, constraints)
```

Comment:
It would be really nice to wire up the errors through the call stack. I think if you are asked to validate a message whose options has a completely incompatible extension with the same number as protovalidate's extension(s), an error is suitable.

Reply:
Agreed, but it'll need to be made as a breaking change (the interface here is exposed indirectly in the public API). Will open an issue.